File: SinaWeibo/ViewControllers/UserInfoViewController.m
//
//  UserInfoViewController.m
//  SinaWeibo
//
//  Created by cxjwin on 13-9-25.
//  Copyright (c) 2013年 cxjwin. All rights reserved.
//

#import <SDWebImage/SDWebImageManager.h>
#import "UserInfoViewController.h"
#import "SinaWeiboManager.h"

@interface UserInfoViewController ()

@end

@implementation UserInfoViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
	// Do any additional setup after loading the view.
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    __weak UIImageView *imageView = (id)[self.view viewWithTag:101];
    
    SinaWeiboCommunicator *communicator = [[SinaWeiboManager defaultManager] communicator];
    NSURLCache *URLCache = communicator.defaultSession.configuration.URLCache;
    
    NSURL *avatarUrl = nil;
    if (self.user.profileImageUrl) {
        if (self.user.avatarLarge) {
            avatarUrl = [NSURL URLWithString:self.user.avatarLarge];
        } else {
            avatarUrl = [NSURL URLWithString:self.user.profileImageUrl];
        }
    }
    
    if (avatarUrl) {
        NSURLRequest *request = [NSURLRequest requestWithURL:avatarUrl];
        NSCachedURLResponse *response = [URLCache cachedResponseForRequest:request];
        if (response) {
            dispatch_async(dispatch_get_main_queue(), ^{
                imageView.image = [UIImage imageWithData:response.data];
            });
        } else {
            [communicator downloadImageWithURL:avatarUrl 
                     downloadCompletionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                         if (error == nil && [data length] > 0) {
                             dispatch_async(dispatch_get_main_queue(), ^{
                                 imageView.image = [UIImage imageWithData:data];
                             });
                         }
                     }];
        }
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    NSLog(@"%s", __func__);
}

@end

File: SinaWeibo/ViewControllers/UserInfoViewController.swift
//
//  UserInfoViewController.swift
//  SinaWeibo
//
//  Created by cxjwin on 14-6-11.
//  Copyright (c) 2014 cxjwin. All rights reserved.
//

import UIKit

class UserInfoViewController: UIViewController {
    
    let communicator = SinaWeiboManager.defaultManager().communicator
    
    var user: WeiboUser?
    
    @IBOutlet var imageView : UIImageView
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Do any additional setup after loading the view.
        self.view.backgroundColor = UIColor.whiteColor()
        
        self.view.addSubview(self.imageView)
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        
        if !self.user {
            return
        }
        
        let curUser = self.user!
        
        var avatarUrl: NSURL?
        
        if curUser.profileImageUrl {
            if curUser.avatarLarge {
                avatarUrl = NSURL(string: curUser.avatarLarge)
            } else {
                avatarUrl = NSURL(string: curUser.profileImageUrl)
            }
        }
        
        if var url = avatarUrl {
            var request = NSURLRequest(URL: url)
            
            var URLCache: NSURLCache? = self.communicator.defaultSession.configuration.URLCache
            
            var response:NSCachedURLResponse? = URLCache?.cachedResponseForRequest(request)
            
            if response {
                var data:NSData? = response!.data
                if data {
                    var image = UIImage(data: data)
                    println("\(image)")
                    self.imageView.image = image
                }
            } else {
                self.communicator.downloadImageWithURL(url, downloadCompletionHandler: {
                    [weak self] (data: NSData?, _, _) in
                    if data?.length > 0 {
                        NSOperationQueue.mainQueue().addOperationWithBlock{
                            var image = UIImage(data: data)
                            println("\(image)")
                            self!.imageView.image = image
                        }
                    }
                    })
            }
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    /*
    // #pragma mark - Navigation
    
    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue?, sender: AnyObject?) {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
    }
    */
    
}
